### PR TITLE
fix: allow 2 threads for CreateIoCompletionPort on single-core to prevent busy looping

### PR DESCRIPTION
**Pull request by bughit**
#### Description of Change

Electron on single-core windows has a problem with cpu pegging when it should be idle, triggered by (among other things) async io.

- electron/electron#12026
- electron/electron#11699
- electron/electron#14319

There are related issues in downstream projects like vscode.

This fix replaces the I/O completion port created by libuv with one with 2 concurrent threads, on single core.

Original idea came from #12026.
@refack recommended to do the iocp override in electron code.
CCing @zcbenz for review

#### Checklist

- [ ] PR description included and stakeholders cc'd
- [ ] `npm test` passes
- [ ] tests are [changed or added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [ ] relevant documentation is changed or added
- [ ] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)

#### Release Notes

Notes: Fixes cpu pegging on single-core windows.

### CODE REVIEW

File: atom/common/node_bindings_win.cc [hunk 1]
```diff
@@ -11,7 +11,16 @@
 namespace atom {
 
 NodeBindingsWin::NodeBindingsWin(BrowserEnvironment browser_env)
-    : NodeBindings(browser_env) {}
+    : NodeBindings(browser_env) {
+  // on single-core the io comp port NumberOfConcurrentThreads needs to be 2
+  // to avoid cpu pegging likely caused by a busy loop in PollEvents
+  char* numProcessors = getenv("NUMBER_OF_PROCESSORS");
```
Comment (nornagon) [inline atom/common/node_bindings_win.cc]: This isn't the right way to test whether we're single-core. Use `base::SysInfo::NumberOfProcessors()`, declared in `base/sys_info.h`.

Comment (refack) [reply atom/common/node_bindings_win.cc]: But it makes for easy testing ;)

**Comment by bughit** [comment]
updated with base::SysInfo::NumberOfProcessors and CHECKS

**Comment by bughit** [comment]
@refack thanks for the help with this

**Comment by refack** [comment]
> masking an underlying issue with some resource being accessed on multiple threads that should only be accessed on a single thread

I'm still investigation that... But I'm not too concerned since I saw the `NodeBindingsWin::PollEvents()` implementation.

**Comment by bughit** [comment]
@zcbenz

- Can this be merged? Is there something I need to do? I think ci failures are not related to this change.
- Since this is a bug fix, is backporting to 4-0-x allowed?

**Comment by zcbenz** [comment]
If no opposition, I would like to merge this PR because:
1. it only affects single-core machines;
2. even if something wrong happens, it won't make things worse since Electron is already very broken on single-core machines with the bug.

**Comment by brenca** [comment]
Just out of curiosity, would this bug repro in node? Could this be eventually upstreamed?

**Comment by bughit** [comment]
> Just out of curiosity, would this bug repro in node?

https://github.com/electron/electron/issues/14319#issuecomment-445079579

> Could this be eventually upstreamed?

https://github.com/electron/electron/issues/14319#issuecomment-445233411
https://github.com/libuv/libuv/pull/2106#issuecomment-445565889

**Comment by bughit** [comment]
3-0-x?

**Comment by bughit** [comment]
@zcbenz Why not 3-0-x?  You are still doing 3.0 releases, doesn't make sense that it would be fixed in 2.0 and not 3.0.

**Comment by BinaryMuse** [comment]
@bughit I believe this was an oversight; I'll get this into the next 3-0-x release.

**Pull request merged**